File: app/models/ativo.rb
class Ativo < ApplicationRecord
  validates :codigo, presence: true, uniqueness: true
  validates :nome, presence: true
  validates :descricao_nota, presence: true
  validates :preco_medio, numericality: true, allow_nil: true
  validates :fator_preco_teto, numericality: true, allow_nil: true
  validates :preco_teto_medio, numericality: true, allow_nil: true
  validates :preco_teto_projetivo, numericality: true, allow_nil: true

  belongs_to :tipo_ativo, required: true
  has_many :operacoes, foreign_key: :ativo_id
  has_many :eventos, foreign_key: :ativo_id
  has_many :posicoes_fim_ano, foreign_key: :ativo_id

  def novo_preco_medio(montante_investido, preco)
    montante_investido = montante_investido.gsub(',', '.').to_f if montante_investido.is_a? String
    preco = preco.gsub(',', '.').to_f if preco.is_a? String

    quantidade = montante_investido / preco

    novo_preco = ((posicao * preco_medio) + montante_investido) / (quantidade + posicao)

    { 'preco_medio_antigo' => preco_medio,
      'preco_medio_novo' => novo_preco.round(2) }
  end

  # rails_admin do
  #   list do
  #     fields :id, :codigo, :nome, :cnpj, :cnpj_fonte_pagadora, :tipo_ativo,
  #            :preco_teto_medio, :preco_teto_projetivo
  #   end
  # end

  def to_param
    codigo
  end

  def preco
    conn = Faraday.new(
      url: 'https://brapi.dev',
      params: { token: ENV.fetch('BRAPI_DEV_TOKEN') }
    )

    response = conn.get("/api/quote/#{codigo}")
    JSON.parse(response.body)['results'].first['regularMarketPrice']
  end
end

File: Stocks/Models/Ativo.cs
namespace Stocks.Models;

public partial class Ativo
{
    public int Id { get; set; }

    public string Codigo { get; set; }

    public string Nome { get; set; }

    public string DescricaoNota { get; set; }

    public string? TipoAcao { get; set; }

    public decimal? PrecoMedio { get; set; }

    public int? Posicao { get; set; }

    public string Cnpj { get; set; }

    public string? CnpjFontePagadora { get; set; }

    public decimal? FatorPrecoTeto { get; set; }

    public decimal? PrecoTetoMedio { get; set; }

    public decimal? PrecoTetoProjetivo { get; set; }

    public int TipoAtivoId { get; set; }

    public virtual ICollection<Evento> Eventos { get; set; } = [];

    public virtual ICollection<Operacao> Operacoes { get; set; } = [];

    public virtual ICollection<PosicaoFimAno> PosicoesFimAno { get; set; } = [];

    public virtual ICollection<Provento> Proventos { get; set; } = [];

    public virtual TipoAtivo TipoAtivo { get; set; }
}
